### ISSUE TITLE
[Bug] Option `--public-path` doesn't work when using esbuild for building libraries

### ISSUE THREAD

**Issue by brillout** [evanw/esbuild#2981]
As the author of a library, I'd like to use esbuild to build the following:

```js
import iconCalendar from './icons/calendar.svg'
import timeUtils from './utils/time.js'

/* ... */
```

Into:

```js
var iconCalendar = '/assets/icons/calendar-RT6VWYUT.svg';
import timeUtils from './utils/time-3GRCYC2A.js'; // Relative path
```

But, if I use `--public-path=/assets/`, I get this instead:

```js
var iconCalendar = '/assets/icons/calendar-RT6VWYUT.svg';
import timeUtils from '/assets/utils/time-3GRCYC2A.js'; // Absolute path
```

Which breaks my library.

I understand the current behavior works for end-user apps, since `utils/time-3GRCYC2A.js` is indeed served at `/assets/utils/time-3GRCYC2A.js`. But that isn't the case for libraries, because the user of the library may use a tool like Next.js that won't be able to process `/assets/utils/time-3GRCYC2A.js`. My users's tools (Next.js, Vite, Parcel, ...) need the relative JavaScript paths to be preserved.

In a nutshell: if the target is the browser, then the current behavior makes sense. But, if the target is the user app (and therefore Next.js, Vite, Parcel, ...), then the current behavior is broken.

**Comment by evanw** [comment]
Sorry, I'm confused. You didn't provide a way to reproduce your issue. How are you generating the code in your post? Marking this issue as `unactionable` because it doesn't have reproduction instructions.

**Comment by brillout** [comment]
A reproduction: https://github.com/brillout/esbuild-reprod-2707. This is the same root cause: only static asset import paths should be transpiled to absolute paths, while JavaScript import paths should be kept relative.

Related feature request: https://github.com/evanw/esbuild/issues/2707. It's a feature request for addressing this issue.

**Comment by evanw** [comment]
Thanks. I can observe the same behavior as you. I put the repro into the playground here, for my reference: [(link)](https://esbuild.github.io/try/#YgAwLjIxLjUAewogIHB1YmxpY1BhdGg6ICcvYXNzZXRzLycsIC8vIFJlbW92ZSB0aGlzIGxpbmUgdG8gZml4IGVycm9yCiAgZW50cnlQb2ludHM6IFsnLi9zcmMvcmVuZGVySHRtbC5qcyddLAogIG91dGRpcjogJy4vZGlzdC8nLAogIGZvcm1hdDogJ2VzbScsCiAgdGFyZ2V0OiAnZXMyMDIwJywKICBidW5kbGU6IHRydWUsCiAgc3BsaXR0aW5nOiB0cnVlLAogIGxvYWRlcjogewogICAgJy5zdmcnOiAnZmlsZScKICB9Cn0AAHNyYy9yZW5kZXJIdG1sLmpzAGltcG9ydCBsb2dvVXJsIGZyb20gJy4vbG9nby5zdmcnCgpyZW5kZXJIdG1sKCkKCmFzeW5jIGZ1bmN0aW9uIHJlbmRlckh0bWwoKSB7CiAgY29uc3QgeyBQYWdlIH0gPSBhd2FpdCBpbXBvcnQoJy4vUGFnZS5qcycpCgogIGNvbnN0IGh0bWwgPSBgPCFET0NUWVBFIGh0bWw+CjxodG1sPgogIDxoZWFkPgogICAgPGxpbmsgcmVsPSJpY29uIiBocmVmPSIke2xvZ29Vcmx9IiAvPgogIDwvaGVhZD4KICA8Ym9keT4KICAgICR7UGFnZSgpfQogIDwvYm9keT4KPC9odG1sPmAKCiAgY29uc29sZS5sb2coaHRtbCkKfQoAAHNyYy9sb2dvLnN2ZwA8c3ZnLz4AAHNyYy9QYWdlLmpzAGV4cG9ydCBmdW5jdGlvbiBQYWdlKCkgewogIHJldHVybiAnSGVsbG8nCn0).

I guess I still don't understand what you're doing. You're writing a library that is going to be bundled by users using a tool of their choice, which could all be wildly different and which understandably needs relative paths to work. But then it seems like you want to use `--public-path` for your assets instead of relative URLs, even though I'd think that same logic would apply to assets as well (the user's tools needing relative paths instead of absolute paths, since they aren't themselves hosting your stuff at specific absolute paths). Why are you using `--public-path` in this situation (for a library and not for an end-user app)?

**Comment by brillout** [comment]
Yes, without SSR, it isn't an issue because the user's bundler (Vite/webpack/...) transpiles the library code (i.e. the code living at `/node_modules/my-library/**/*`).

But, with SSR and Vite, library code isn't transpiled: by default Vite externalizes dependencies on the server-side. This means that the library code (`/node_modules/my-library/**/*`) is directly executed by Node.js (without any Vite transpilation).

In that scenario, it's a good thing that asset imports are transpiled:

```diff
- import iconCalendar from './icons/calendar.svg'
+ var iconCalendar = '/assets/icons/calendar-RT6VWYUT.svg';
```
Because Node.js cannot import `.svg` files and would crash otherwise.

But for JavaScript imports, it's a bad thing as the following makes Node.js crash:

```diff
- import timeUtils from './utils/time.js'
+ import timeUtils from '/assets/utils/time-3GRCYC2A.js';
```

Node.js cannot import `/assets/utils/time-3GRCYC2A.js` and crashes.

Thus, the requirement of only setting the public base for asset imports while leaving JavaScript imports untouched.

**Comment by hyrious** [comment]
I understand that you're looking for a way to _unify_ the code style to reference assets in js modules for both browser and Node.js to load. The problem here is that there's no such behavior specified in JavaScript language, nor the same API to use in every environment. `import('/path/to/file.svg')` is also not supported by browsers.

Your proposed solution was to only prepend absolute paths to assets. However that won't solve your initial problem of making your library work in Vite SSR build since the absolute path does not actually point to the file when resolved from the root of the Vite project:

```jsx
// /node_modules/your-lib/index.js
var file = '/assets/file.svg'
export var Icon = h('img', { src: file }, null)

// /src/index.jsx
import {Icon} from 'your-lib'
render(<Icon />, document.body)
```

Note that the actuall `file.svg` is at `/node_modules/your-lib/assets/file.svg`. Result:

```html
<img src="/assets/file.svg">  <!-- 404 -->
```

This is because when the file becomes a string in your js module, it loses the information that it links an external file. No bundler could recognize that fact and copy the file to the correct location.

- - -

There're some other actions you may take:

- For SVG files, you can bundle them in your js modules with `--loader:.svg=dataurl`. So there will be no external asset files.
  - If you are working with JSX, you can embed you SVG elements in JSX directly.
- If you're dealing with some large assets like wasm files, where it is better to load them separatedly. You can construct or require the URL to the asset. For example esbuild-wasm requires you to run [`esbuild.initialize({ wasmURL })`](https://esbuild.github.io/api/#browser) to help it find the file.
